Mainnet/HashBattle/HashBattle/Arena.cs

```diff
@@ -0,0 +1,194 @@
+﻿using Stratis.SmartContracts;
+using System;
+using System.Text;
+
+/// <summary>
+/// A Stratis smart contract for running a game battle where owner will start the battle and maximum 4 users can enter a battle
+/// </summary>
+public class Arena : SmartContract
+{
+    public Arena(ISmartContractState smartContractState)
+    : base(smartContractState)
+    {
+        BattleOwner = Message.Sender;
+    }
+
+    /// <summary>
+    /// Set the address deploying the contract as battle owner
+    /// </summary>
+    private Address BattleOwner
+    {
+        get => PersistentState.GetAddress(nameof(BattleOwner));
+        set => PersistentState.SetAddress(nameof(BattleOwner), value);
+    }
+
+    /// <summary>
+    /// Battle owner will start the battle
+    /// </summary>
+    public bool StartBattle(ulong battleId, ulong fee)
+    {
+        Assert(Message.Sender == BattleOwner, "Only battle owner can start game.");
+
+        var battle = new BattleMain();
+        battle.BattleId = battleId;
+        battle.MaxUsers = 4;
+        battle.Fee = fee;
+        battle.Users = new Address[battle.MaxUsers];
+        SetBattle(battleId, battle);
+
+        Log(battle);
+        return true;
+    }
+
+    /// <summary>
+    /// 4 different user will enter the battle
+    /// </summary>
+    public bool EnterBattle(ulong battleId, uint userindex)
+    {
+        var battle = GetBattle(battleId);
+
+        Assert(battle.Winner == Address.Zero, "Battle ended.");
+
+        Assert(battle.Fee == Message.Value, "Battle amount is not matching.");
+
+        var user = GetUser(battleId, Message.Sender);
+
+        Assert(!user.ScoreSubmitted, "The user already submitted score.");
+
+        user.Address = Message.Sender;
+
+        SetUser(battleId, Message.Sender, user);
+
+        battle.Users.SetValue(user.Address, userindex);
+        SetBattle(battleId, battle);
+
+        Log(battle);
+        return true;
+    }
+
+    /// <summary>
+    /// 4 different user will end the battle and submit the score
+    /// </summary>
+    public bool EndBattle(Address userAddress, ulong battleId, uint score, bool IsBattleOver)
+    {
+        Assert(Message.Sender == BattleOwner, "Only battle owner can end game.");
+
+        var battle = GetBattle(battleId);
+
+        Assert(battle.Winner == Address.Zero, "Battle ended.");
+
+        var user = GetUser(battleId, userAddress);
+
+        Assert(!user.ScoreSubmitted, "The user already submitted score.");
+
+        user.Score = score;
+        user.ScoreSubmitted = true;
+
+        SetUser(battleId, userAddress, user);
+
+        if (IsBattleOver)
+            ProcessWinner(battle);
+
+        Log(user);
+        return true;
+    }
+
+    /// <summary>
+    /// Get winner address
+    /// </summary>
+    public Address GetWinner(ulong battleId)
+    {
+        var battle = GetBattle(battleId);
+        Log(battle);
```

Comment:
This log serves no purpose and should be removed. Logs should only be used for events, or in other words logging state changes, otherwise you're wasting gas and making it a lot harder to determine what happened on chain.